jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java

```diff
@@ -90,17 +90,20 @@ public static WebSocketComponents ensureWebSocketComponents(Server server, Servl
         if (server.contains(bufferPool))
             serverComponents.unmanage(bufferPool);
 
-        servletContext.setAttribute(WEBSOCKET_COMPONENTS_ATTRIBUTE, serverComponents);
-        LifeCycle.start(serverComponents);
-        servletContext.addListener(new ServletContextListener()
+        // Stop the WebSocketComponents when the ContextHandler stops.
+        ContextHandler contextHandler = Objects.requireNonNull(ContextHandler.getContextHandler(servletContext));
+        contextHandler.addManaged(serverComponents);
+        contextHandler.addEventListener(new LifeCycle.Listener()
         {
             @Override
-            public void contextDestroyed(ServletContextEvent sce)
+            public void lifeCycleStopping(LifeCycle event)
             {
-                LifeCycle.stop(serverComponents);
+                contextHandler.removeBean(serverComponents);
```

Comment:
Do you need to stop the serverComponents?

Reply:
Yes, the `removeBean()` call will stop the `serverComponents`.